### ISSUE TITLE
Poor colour choice and column lables for pub outdated versions.

### ISSUE THREAD

**Issue by bsutton** [dart-lang/pub#2611]
I'm running dart 2.9 and using `pub outdated`

I find the output from pub outdated less than intuitive.

First problem is the colour scheme.
The colour red is broadly used by the community to indicate a problem.

pub outdated seems to be using red to mean the opposite of the broader conventions.

To clarify, I see that pub outdated's job is to 'highlight' upgrade opportunities hence 'red' indicating the opposite to my expectations.

Its also unclear why the 'current' column is colour coded. Surely current is always current so it seems to make no senses why this is red.

The colour coding should only be applied if there is a useful action to be taken or advice to be given.

The second problem is that I don't find the column names very useful.
In particular even after reading the doco the terms 'upgradable' and 'resolvable' are still clear as mud.

I think a better label for 'upgradable' is  'Constrained'.

The doco for Upgradable states:
> The latest version allowed by your pubspec.yaml file. 

This is clearly talking about a constraint.

The term 'Resolvable' requires the user to be familiar with the concept of package resolution and what a resolvable packing implies. This level of understanding should not be a pre-requisite for successful use of pub upgrade.

I think a better team is 'Reachable'

> Other options for 'Resolvable' are 'Available' or 'Upgradable'.

The documentation would then read as follows (which I believe is much clearer).

```
#Constrained
The latest version allowed by your pubspec.yaml file. This is the maximum version that pub upgrade 
will upgrade to given the current version constraints in your pubspec.yaml. 
The value is - if the value in the Current column is -.

#Reachable
The maximum version that the package can be upgrade to, when combined with all other dependencies. 
This version corresponds to what pub upgrade gives you if all version constraints in pubspec.yaml are unbounded. 
A value of - means that the package won’t be needed.
```
Note: I use the term 'maxiumum' rather than 'latest' as it is more evocative when talking about numbered version constraints.

So in summary

Labels
'Upgradable' becomes 'Constrained'
'Resolvable' becomes 'Reachable'

Colours
Current - this should always be white text as there is nothing of particular interest here.
Upgradable/Constrained should be 'green' IF the package can be upgraded or white if no upgrade is available.
Resolvable/Reachable should be white if no actions can be taken, but green if the version can be reach once constraints are removed.
Latest should always be white.

**Comment by bsutton** [comment]
I've just had another look at the current colour coding and it would appear that 'red' is currently used to indicate that 'no action' is required. This is actually worse that I thought as we are highlighting columns where no action is required.

This gets worse as it appears that the 'latest' column uses 'grey' to indicate no action can be applied. So the ui uses both 'red' and 'grey' to indicate no action can be applied.

The output then uses white to indicate action is required, but of course the headings also use white.

Colour should be used consistently and each colour should have a distinctive meaning. This helps guide the users eye quickly to the main points where action is required.

**Comment by sigurdm** [comment]
Thanks for the feedback.

Re: Colors:

> I've just had another look at the current colour coding and it would appear that 'red' is currently used to indicate that 'no action' is required.

That is not true: Red is used to indicate "not the latest" white is used to indicate "latest" and grey is used to indicate "same version as previous column - nothing to see here".

There is a --no-color option that uses `*` for "not the latest" and no marking for the latest.

Does this explanation make sense?

CC @mit-mit you came up with the grey color if I remember. Do you have any comments?

Re: Labels:

We spend quite some time deciding what to call these columns and I still think that they do work quite well.

I think your suggestions are also fine, but I am not sure they are much better - I think it would be breaking to change them at this point (they are used in the json output).

@kwalrath Do you have any input?

**Comment by mit-mit** [comment]
Yes, red was meant to indicate where an action should be taken and grey that an action should not or cannot be taken. I think that is how it works currently?

I personally find 'Constrained' harder to understand than 'Upgradeble'. With the latter, there is the nice. consistency that the column lists what `pub upgrade` will give you.

**Comment by bsutton** [comment]
@mit-mit 
> Yes, red was meant to indicate where an action should be taken and grey that an action should not or cannot be taken. I think that is how it works currently?

No that isn't the case.
Currently red appears to indicate that no action is required.
After updating all packages to the upgradable state I'm presented with a sea of red.

If 'red' is meant to indicate an action, then why would 'current' ever be red. The very presence of a package in the list tells me that some attention is needed so highlighting the current version provides no additional information and if facts distracts me from the important info.

![image](https://user-images.githubusercontent.com/900408/89961613-96fac780-dc85-11ea-9222-cd45f5de5eed.png)

Look at the line for coverage. My understanding is that this indicates that I'm on the latest version possible, yet I get three red indicators. Clearly red doesn't mean an action is required.

The analyzer line appears to be working as per your above notes, so maybe there is a bug in the colour selection. However it makes no sense for the 'current' column to be red. The 'Upgradable' column should be red so that the UI takes you directly to where and what action can be taken. Highlighting the 'upgradable' column in red tells me that an upgrade is recommended and what version it will take me to.  Highlighting the 'current' column provides no useful information.

The pubspec from the above example:
```
name: dshell
version: 1.10.18
homepage: https://github.com/bsutton/dshell
description: An API and dev environment for writing console apps/scripts using dart. e.g. allows you to build bash style scripts with dart.
environment: 
  sdk: '>=2.6.0 <3.0.0'
dependencies: 
  archive: ^2.0.13
  args: ^1.5.2
  collection: ^1.14.12
  csv: ^4.0.3
  equatable: ^1.0.1
  file: ^5.1.0
  file_utils: ^0.1.3
  glob: ^1.2.0
  ini: ^2.0.1
  intl: ^0.16.0
  logger: ^0.9.1
  meta: ^1.1.8
  money2: ^1.4.0
  path: ^1.7.0
  pedantic: ^1.6.0
  pubspec: ^0.1.3
  pub_semver: ^1.4.2
  system_info: ^0.1.3
  uuid: ^2.0.4
  validators: ^2.0.0+1
  yaml: ^2.2.0
dev_dependencies: 
  test: ^1.6.0
  effective_dart: ^1.0.0
  mockito: ^4.0.0
  test_coverage: ^0.4.2
executables: 
  dshell: dshell
  dshell_complete: dshell_complete
  dshell_install: dshell_install
```

Here is another example were I'm getting a sea of red when in reality there is no actions that can be taken (ignoring the first line).
![image](https://user-images.githubusercontent.com/900408/89967301-7b4aed80-dc94-11ea-9a75-13f9287b4223.png)


Again changing the 'current' column to red provides no information and is a visual distraction (why are you trying to draw my eye to a column that requires no action).


Taking on board your statements, my new summary would be:

Current - always grey 
Upgradable 
- red if an upgrade can occur by removing a constraint.
- grey if its the same value as current.

Resolvable 
- red if an upgrade can occur without introducing conflicts.
- grey if the version is the same as the prior column

Latest
- grey if same as resolvable
- white if the version no. is different to resolvable.

I concur that the breaking change may be an issue so the middle ground is probably improving the doco as I've suggested above but continue to use the term 'Upgradable' and 'Reachable'.

**Comment by sigurdm** [comment]
Red is not meant to mean action, but a version that is not the newest available. 
Often that indicates that some action is desirable - but there might be mutual incompatibilities resulting in no possible action (other than using other packages).

Closing as WONTFIX.
